nextstrain/cli/remote/s3.py

```diff
@@ -78,33 +43,128 @@ def upload(local_files: List[Path], bucket, prefix: str) -> List[str]:
     files = list(zip(local_files, [ prefix + f.name for f in local_files ]))
 
     for local_file, remote_file in files:
-        print("Deploying", local_file, "as", remote_file)
+        yield local_file, Path(remote_file)
 
         # Upload compressed data
-        with local_file.open("rb") as data, gzip_stream(data) as gzdata:
+        with GzipCompressingReader(local_file.open("rb")) as gzdata:
             bucket.upload_fileobj(
                 gzdata,
                 remote_file,
                 { "ContentType": content_type(local_file), "ContentEncoding": "gzip" })
 
-    return [ remote for local, remote in files ]
+    # Purge any CloudFront caches for this bucket
+    purge_cloudfront(bucket, [remote for local, remote in files])
+
+
+def download(url: urllib.parse.ParseResult, local_path: Path, recursively: bool = False) -> Iterable[Tuple[Path, Path]]:
+    """
+    Download the files deployed at the given remote *url*, optionally
+    *recursively*, saving them into the *local_dir*.
+    """
+    bucket, path = split_url(url)
+
+    # Download either all objects sharing a prefix or the sole object (if any)
+    # with the given key.
+    if recursively:
+        objects = [ item.Object() for item in bucket.objects.filter(Prefix = path) ]
+    else:
+        objects = [ bucket.Object(path) ]
+
+    def local_file_path(obj):
+        if local_path.is_dir():
+            return local_path / Path(obj.key).name
+        else:
+            return local_path
+
+    files = list(zip(objects, [local_file_path(obj) for obj in objects]))
+
+    for remote_object, local_file in files:
+        yield Path(remote_object.key), local_file
+
+        encoding = remote_object.content_encoding
```

Comment:
Downloading a file which doesn't exist (which I'll do a lot as I'll be meaning to use `--recursive` but forget the flag) results in a trace from this line. This should be caught & help printed. 

```
Traceback (most recent call last):
  File "/Users/naboo/miniconda3/envs/nextstrain-cli/bin/nextstrain", line 11, in <module>
    load_entry_point('nextstrain-cli', 'console_scripts', 'nextstrain')()
  File "/Users/naboo/github/nextstrain/cli/nextstrain/cli/__main__.py", line 10, in main
    return cli.run( argv[1:] )
  File "/Users/naboo/github/nextstrain/cli/nextstrain/cli/__init__.py", line 56, in run
    return opts.__command__.run(opts)
  File "/Users/naboo/github/nextstrain/cli/nextstrain/cli/command/remote/download.py", line 73, in run
    for remote_file, local_file in downloads:
  File "/Users/naboo/github/nextstrain/cli/nextstrain/cli/remote/s3.py", line 84, in download
    encoding = remote_object.content_encoding
  File "/Users/naboo/miniconda3/envs/nextstrain-cli/lib/python3.7/site-packages/boto3/resources/factory.py", line 339, in property_loader
    self.load()
  File "/Users/naboo/miniconda3/envs/nextstrain-cli/lib/python3.7/site-packages/boto3/resources/factory.py", line 505, in do_action
    response = action(self, *args, **kwargs)
  File "/Users/naboo/miniconda3/envs/nextstrain-cli/lib/python3.7/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(**params)
  File "/Users/naboo/miniconda3/envs/nextstrain-cli/lib/python3.7/site-packages/botocore/client.py", line 320, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/naboo/miniconda3/envs/nextstrain-cli/lib/python3.7/site-packages/botocore/client.py", line 623, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (404) when calling the HeadObject operation: Not Found
```